### ISSUE TITLE
Not able to integrate in Swift

### ISSUE THREAD

**Issue by talvinder** [BranchMetrics/Branch-iOS-Invite-SDK#23]
I have been trying hard to follow the documentation, but facing tough issues.

Check out the screenshot
![screenshot 2015-05-31 19 14 56](https://cloud.githubusercontent.com/assets/6104304/7902115/76718be8-07c9-11e5-8f1b-01380b5b2c8d.png)

I just copy pasted the given code and got these errors. Tried to follow what xcode suggested which led to the error message shown in the screenshot.

Please suggest solutions.

**Comment by shortstuffsushi** [comment]
Hi @talvinder. Sorry about that. I'm not real familiar with swift, but it appears that your `AppDelegate` isn't of valid type, likely because it doesn't implement the `BranchWelcomeControllerDelegate` protocol. Is that true? If not, I will have to dig more.

**Comment by talvinder** [comment]
I have added ````BranchWelcomeControllerDelegate```` in ````AppDelegate```` but to no avail.

![screenshot 2015-05-31 20 44 04](https://cloud.githubusercontent.com/assets/6104304/7902387/f392b49c-07d5-11e5-8699-de1b5367500c.png)

**Comment by shortstuffsushi** [comment]
The second argument is also supposed to be a dictionary, so NSObject anyObject may not work. 

But again, not familiar with Swift. I will try to look this afternoon.

**Comment by talvinder** [comment]
Hi @shortstuffsushi were you able to look into it?

I typecasted the second argument as NSDictionary but that didn't help either.

**Comment by shortstuffsushi** [comment]
Hi again @talvinder. I'm sorry, I didn't get around to this last night. I am looking now, though!

**Comment by shortstuffsushi** [comment]
Ok @talvinder, I think I found the problem.

You'll notice if you start to type in `BranchWelcomeViewController.branchWelcome....`, Xcode has no suggestions, as though these methods don't exist! Apparently, this is on purpose. According to [this SO post](http://stackoverflow.com/questions/24295476/swift-can-not-use-nsimage-imagenamed),

>The compiler automatically remaps ObjC class methods that are named as convenience constructors to work as Swift initializers.

So, if you just use
```
let branchWelcomeController:BranchWelcomeViewController = BranchWelcomeViewController(delegate:self, branchOpts:params)
```

It should work for you! Let me know if that helps.

**Comment by talvinder** [comment]
Hey thanks for responding.

But the error stays, see the screenshot please

![screenshot 2015-06-01 21 20 27](https://cloud.githubusercontent.com/assets/6104304/7916878/40709788-08a4-11e5-9b81-4e598302986b.png)

**Comment by shortstuffsushi** [comment]
I think at this point, and based on the screrenshot you took above, that's occurring because your AppDelegate doesn't implement the delegate methods correctly. At minimum, it needs to include `func welcomeControllerConfirmedInvite()` and `func welcomeControllerWasCanceled()`. For me, with those two methods implemented, this compiled :+1:

**Comment by talvinder** [comment]
Thanks @shortstuffsushi it compiled!

Had to add the delegate methods plus added the BranchWelcomeControllerDelegate to appdelegate class. I believe this info should be added to the readme section, will be valuable for other folks. Will be happy to contribute

**Comment by talvinder** [comment]
The joy was short lived. Now I am stuck in invite button pressed method.

Any suggestions?
I tried adding BranchInviteControllerDelegate to the class declaration, but swift didn't show any suggestions and as expected the error was thrown. In the swift suggestions only BranchWelcome.... suggestions are present. I installed using cocoapods.

Screenshot of xcode suggestions:
![screenshot 2015-06-01 21 59 59](https://cloud.githubusercontent.com/assets/6104304/7917672/97b77e8a-08a9-11e5-8ca2-fb13d690b11a.png)


Check the screenshot of error:
![screenshot 2015-06-01 21 55 20](https://cloud.githubusercontent.com/assets/6104304/7917575/02c19c52-08a9-11e5-8cf0-9ceb9809427e.png)

**Comment by shortstuffsushi** [comment]
@talvinder have you imported that class in your bridging header?

**Comment by shortstuffsushi** [comment]
Actually, there isn't a convenience initializer for that one, as the invite currently returns a `UINavigationController`. You can call `BranchInviteViewController.branchInviteViewControllerWithDelegate` just fine, though. I'm going to close this for now, if you still have trouble, feel free to reopen.

**Comment by talvinder** [comment]
hi @shortstuffsushi , sorry but I am facing more troubles. Check out the screenshots below:

I did the imports
![screenshot 2015-06-02 01 47 36](https://cloud.githubusercontent.com/assets/6104304/7922420/c991029e-08c9-11e5-877f-c741984a91c2.png)

I declared the delegates and added functions to conform, still facing errors
![screenshot 2015-06-02 01 49 50](https://cloud.githubusercontent.com/assets/6104304/7922441/eac8089a-08c9-11e5-9936-8bf1615eca1f.png)

**Comment by shortstuffsushi** [comment]
@talvinder the error you're seeing tells you exactly what the problem is -- you're not implementing `BranchInviteControllerDelegate`. You need to add the required methods from that protocol.

**Comment by talvinder** [comment]
I added all the required methods:

inviteControllerDidFinish
inviteControllerDidCancel
invitingUserFullname

I ignored the optional methods.

**Comment by shortstuffsushi** [comment]
Could you maybe link to a gist containing the contents of that file? From here with the screenshots you're sending, it's difficult to see what you're missing. That error means that some required method isn't present, though.

**Comment by talvinder** [comment]
hey @shortstuffsushi 
firstly thanks for being so prompt.

here is the gist: https://gist.github.com/talvinder/662457791772b887aca6

**Comment by shortstuffsushi** [comment]
[This method](https://gist.github.com/talvinder/662457791772b887aca6#file-fdviewcontroller-L20) must return a string, that's probably why it's not working for you.

**Comment by talvinder** [comment]
Hi @shortstuffsushi 

I got things to work. The controller is launching. Now the issue is with inviteContactProviders

I imported BranchInviteTextContactProvider.h to bridge

No idea how to translate this to swift:

````
- (NSArray *)inviteContactProviders {
    return @[
        // SMS provider
        [BranchInviteTextContactProvider textContactProviderWithInviteMessageFormat:@"Check out my demo app with Branch:\n\n%@!"]

        // Email provider
        //[BranchInviteEmailContactProvider emailContactProviderWithSubject:@"Check out this demo app!" inviteMessageFormat:@"Check out my demo app with Branch:\n\n%@!"],

        // Add your own custom provider
        //[[MysteryIncContactProvider alloc] init]
    ];
}
````

**Comment by shortstuffsushi** [comment]
Hi @talvinder, looks like you're making progress. Unfortunately, I don't have time to walk you through every method of the integration. Could you try to take some time and experiment with it on your own, then contact us if you run into any issues.
